### PR TITLE
Sample mean transcript coverage over a list of genes

### PR DESCRIPTION
**Pull request by northwestwitch**
### This PR adds | fixes:
- Return coverage over a list of genes (transcripts only) for a sample in the database

### How to test:
- Launch an instance of the app and provide a sample with a d4 file and a list of genes in various formats (ensembl ids, hgnc ids, hgnc symbols)

### Expected outcome:
- [x] Mean transcript coverage should be returned for each provided gene

### Review:
- [ ] Code approved by
- [x] Tests executed by CR

This [version](https://semver.org/) is a:
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

**Comment by northwestwitch** [comment]
Local tests with a sample with a real d4 file:

- By providing a list of HGNC symbols: 🆗 

image

- By providing a list of HGNC IDs: 🆗 

image

- By providing a list of Ensembl IDs: 🆗 

image

**Comment by northwestwitch** [comment]
This one is ready for review

File: src/chanjo2/meta/handle_d4.py [hunk 1]
```diff
@@ -58,3 +62,42 @@ def genes_coverage(d4_file: D4File, genes: List[SQLGene]) -> List[CoverageInterv
...
+        mean_gene_cov: float = 0
+        if gene_transcripts:
+            mean_gene_cov: float = mean(
+                d4_file.mean(
+                    [
+                        (transcript.chromosome, transcript.start, transcript.stop)
+                        for transcript in gene_transcripts
+                    ]
+                )
+            )
```
Comment (henrikstranneheim) [inline src/chanjo2/meta/handle_d4.py:90]: I think this part would be good as a function:
mean_gene_coverage: float = get_mean_gene_coverage(args)
Maybe we can make it general enough to reuse for exons.

File: src/chanjo2/models/pydantic_models.py [hunk 2]
```diff
@@ -133,10 +133,10 @@ class CoverageInterval(BaseModel):
     ensembl_gene_id: Optional[str]
     hgnc_id: Optional[int]
     hgnc_symbol: Optional[str]
+    interval_id: Optional[int]
```
Comment (henrikstranneheim) [inline src/chanjo2/models/pydantic_models.py:136]: Would be nice to order this alphabetically to make it easier to scan

File: tests/src/chanjo2/endpoints/test_coverage.py [hunk 3]
```diff
@@ -163,23 +163,23 @@ def test_d4_intervals_coverage(
...
+def test_sample_gene_coverage_hgnc_symbols(
     build: str,
     demo_client: TestClient,
     endpoints: Type,
     genomic_ids_per_build: Dict[str, List],
 ):
     """Test the function that returns the coverage over multiple genes of a sample when a list of HGNC symbols is provided."""
 
-    # GIVING a sample coverage query containing HGNC gene symbols
+    # GIVING a sample gene coverage query containing HGNC gene symbols
```
Comment (henrikstranneheim) [inline tests/src/chanjo2/endpoints/test_coverage.py:174]:
```suggestion
    # GIVEN a sample gene coverage query containing HGNC gene symbols
```
Fix for all occurrences

File: tests/src/chanjo2/endpoints/test_coverage.py [hunk 4]
```diff
@@ -215,23 +215,107 @@ def test_samples_gene_coverage_hgnc_ids(
...
         "ensembl_ids": genomic_ids_per_build[build]["ensembl_gene_ids"],
     }
 
     # THEN the response should be successful
-    response = demo_client.post(endpoints.INTERVALS_SAMPLE_COVERAGE, json=sample_query)
+    response = demo_client.post(endpoints.SAMPLE_GENES_COVERAGE, json=sample_query)
+    assert response.status_code == status.HTTP_200_OK
+
+    # AND return coverage intervals data
+    result = response.json()
```
Comment (henrikstranneheim) [inline tests/src/chanjo2/endpoints/test_coverage.py:238]:
```suggestion
    coverage_intervals: list = response.json()
```

File: tests/src/chanjo2/endpoints/test_coverage.py [hunk 5]
```diff
@@ -215,23 +215,107 @@ def test_samples_gene_coverage_hgnc_ids(
...
+    }
+
+    # THEN the response should be successful
+    response = demo_client.post(
+        endpoints.SAMPLE_TRANSCRIPTS_COVERAGE, json=sample_query
+    )
+    assert response.status_code == status.HTTP_200_OK
+
+    # AND return coverage intervals data
+    result = response.json()
```
Comment (henrikstranneheim) [inline tests/src/chanjo2/endpoints/test_coverage.py:266]: Same here.

File: tests/src/chanjo2/endpoints/test_coverage.py [hunk 6]
```diff
@@ -215,23 +215,107 @@ def test_samples_gene_coverage_hgnc_ids(
...
+
+
+@pytest.mark.parametrize("build", Builds.get_enum_values())
+def test_sample_transcripts_coverage_ensembl_ids(
+    build: str,
+    demo_client: TestClient,
+    endpoints: Type,
+    genomic_ids_per_build: Dict[str, List],
+):
+    """Test the function that returns the coverage over  transcripts of multiple genes of a sample when a list of Enseml IDs is provided."""
```
Comment (henrikstranneheim) [inline tests/src/chanjo2/endpoints/test_coverage.py:306]:
```suggestion
    """Test the function that returns the coverage over transcripts of multiple genes of a sample when a list of Enseml IDs is provided."""
```

**Review by henrikstranneheim — APPROVED** [review]
Fantastic 👍. It is getting more and more complete

File: src/chanjo2/meta/handle_d4.py [hunk 7]
```diff
@@ -58,3 +62,42 @@ def genes_coverage(d4_file: D4File, genes: List[SQLGene]) -> List[CoverageInterv
             )
         )
     return genes_cov
+
+
+def genes_transcript_coverage(db: Session, d4_file: D4File, genes: List[SQLGene]):
```
[the remark this answers is not in the document]
Comment (northwestwitch) [reply src/chanjo2/meta/handle_d4.py]: Done

File: tests/src/chanjo2/endpoints/test_coverage.py (hunk 4 above)
Comment (northwestwitch) [reply tests/src/chanjo2/endpoints/test_coverage.py]: Done

File: src/chanjo2/meta/handle_d4.py (hunk 1 above)
Comment (northwestwitch) [reply src/chanjo2/meta/handle_d4.py]: Done

**Comment by northwestwitch** [comment]
Ok, I think I have addressed all your comments @henrikstranneheim. I would like to integrate the completeness also in the transscripts stats, but I'd rather do it in another PR, so it's less messy!

**Pull request merged**